### PR TITLE
Rework provider and fetching documentation

### PR DESCRIPTION
**Pull request by levitte**
All the current fetching functions were collected in one manual page, together with a very general description of providers, algorithm fetching and the providers we deliver.

Some of that is reference material that does belong in manual section 3, but the more general description is better placed in manual section 7.

This also includes provider-base(7), which describes the base tables of functions that libcrypto and the provider pass to each other, thereby providing a baseline with which they can communicate further with each other.  This also contains an example for a ficticious provider, providing an implement of a fictitious algorithm for a fictitious operation.

### CODE REVIEW

**Comment by levitte** [comment]
This work is incomplete in itself, as provider(7) refers to other pages in section 7 that are yet to be written.  I'm unsure if that should be for this PR or follow up PRs.  Maybe this one is big enough as it is?

**Comment by levitte** [comment]
I've added provider-base(7), showing approximately how i imagine the contents of the dispatch tables being described.

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Typo: often

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Typo: return

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: "closer description" sounds wrong to me. I would see "a more detailed description"

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Typo: used

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Why is Providers capitalised here?

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Two typos: implementations *of* the same *algorithms*

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: KMAC128, KMAC256, MD5

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]:
````
=item Key Exchange

dhKeyAgreement
````

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Two typos: implementations *of* the same *algorithms*

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]:
````
=item Digests

SHA1, SHA224, SHA256, SHA384, SHA512, SHA512-224, SHA512-256, SHA3-224, SHA3-256, SHA3-384, SHA3-512, KMAC128, KMAC256

=item Symmetric ciphers

AES-256-ECB, AES-192-ECB, AES-128-ECB, AES-256-CBC, AES-192-CBC, AES-128-CBC, AES-256-CTR, AES-192-CTR, AES-128-CTR
````

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: *of* the same *algorithms*

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Not MD5 (that's in default)

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Side note: This seems like a very strange function for us to be providing like this

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Typo: pointer*s*

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Typo: *to* indicate

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Two typos: *the* source *of* the error

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Side note: This kind of seems weird. We have a built-in "provider" property that we don't seem to be using for our own providers. Instead we're using "default=yes", "legacy=yes" or "fips=yes". Should we instead by setting "provider=default", "provider=legacy" or "providers=fips" for our built-in stuff?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good. Quite a few comments, but mostly minor typos.

File: doc/man3/EVP_DigestInit.pod
Comment (levitte) [reply doc/man3/EVP_DigestInit.pod]: I agree

File: doc/man3/EVP_DigestInit.pod
Comment (levitte) [reply doc/man3/EVP_DigestInit.pod]: Okie. Makes kinda sense. I assume it's an error to use a fetched method for this

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: I have an ongoing branch for OSSL_STORE support. There, the "algorithm" name is the URI scheme. That's what I have in mind.

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: Head1 titles usually are. The better question is why I didn't capitalise openssl

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: It does memory allocation, that's the reason it's bunched in like this. I kinda [shrug], ya know?

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: Maybe we should

File: doc/man7/provider-base.pod
Comment (mattcaswell) [reply doc/man7/provider-base.pod]: Well...yeah....it calls OPENSSL_malloc...but so do lots of other functions in the library - so why this one? Anyway this was a side note.

Comment (levitte) [reply doc/man7/provider-base.pod]: That à question to ask @richsalz

Comment (richsalz) [reply doc/man7/provider-base.pod]: Poof, summon me and I appear :)  The answer is in the message for commit b60cba3c5d25dcd5cea2c97193a30b730c8a0f28

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: "they return", even, right?

Comment (richsalz) [reply doc/man7/provider.pod]: typo: "oftemn"

Yes, "diverse" is plural so you "they return"

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: "where in the provider source the error occured"...  your suggestion changes it to "where in the provider the source of the error occured", which doesn't make sense to me.

**Comment by levitte** [comment]
Lots of comments addressed

File: doc/man7/provider-base.pod
Comment (richsalz) [reply doc/man7/provider-base.pod]: How about "to indicate where the error occurred or was reported." Using "the source of the error" is bulky. And I think "the provider source" means it's not needed.

Comment (levitte) [reply doc/man7/provider-base.pod]: Good idea

**Comment by levitte** [comment]
I've decided that this set will be enough for *this* PR.  We will still need to document the CIPHER, DIGEST, KEYEXCH and KEYMGMT operations, but I'd like to leave that to other PRs.

(part of this decision is giving others a chance to document as well :wink:)

File: doc/man7/provider-base.pod
Comment (mattcaswell) [reply doc/man7/provider-base.pod]: That doesn't seem to answer the question. Just because EVP control functions need hexstring conversion doesn't seem like a good reason to make this function available to all providers.

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: s/have/has/

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Should we document here what parameters are currently supported?

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: s/objects/object/

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Describe `num` and `args`?

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Do we need to specify which of these are mandatory to implement and which are optional?

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: Should be `void **provctx`

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: We seem to end halfway through a sentence here.

File: doc/man7/provider.pod
Comment (mattcaswell) [reply doc/man7/provider.pod]: Doesn't seem resolved

File: doc/man7/provider.pod
Comment (mattcaswell) [inline doc/man7/provider.pod]: s/FIPS/legacy/

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: That's a discussion for another PR.  This PR documents what's already present.

Comment (mattcaswell) [reply doc/man7/provider-base.pod]: Yes...this was always a side note.

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: No.  That should be documented in the provider documentation.  Exactly how, I do not currently know.  Feel free to expand on that.

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: Good point

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: I agree.  I copied this from...  somewhere.  I'll have to find it again.

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: True.  Will fix

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: Done noe

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: Done

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: (doesn't have to be in *this* PR)

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: Done

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: None of them are actually mandatory.  I've added a small paragraph pointing out that a provider lacking some of these is fairly useless.

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: I have no idea what I meant to write there, so I'm removing that half sentence

File: doc/man7/provider.pod
Comment (levitte) [reply doc/man7/provider.pod]: Done

File: doc/man7/provider-base.pod
Comment (mattcaswell) [reply doc/man7/provider-base.pod]: > No. That should be documented in the provider documentation.

Umm. But this is core functionality? So surely it makes sense to document it here. Currently we only support "openssl-version" and "provider-name" along with any parameters from the config file. What specific config file parameters a provider supports should be in provider specific documentation. But the fact that config file parameters are exposed to the provider through this call should surely be documented here (along with documentation for "openssl-version" and "provider-name")

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Typo: and

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: Good points

File: doc/man7/provider-base.pod
Comment (richsalz) [reply doc/man7/provider-base.pod]: I don't want to hijack this PR, if there's another place to discuss this let me know.  EVP uses hexstring, EVP is in the boundary, therefore hexstring needs to be in the boundary as well. I remain confused by Matt's concern.

Comment (levitte) [reply doc/man7/provider-base.pod]: There's no real need to export OPENSSL_hexstr2buf(), the providers that use EVP could use the source of that function directly, as long as they make sure to have OPENSSL_malloc() and friends available within their boundaries *somehow*

**Comment by levitte** [comment]
Side discussion aside, where are we with this?

File: doc/man7/provider-base.pod
Comment (mattcaswell) [inline doc/man7/provider-base.pod]: Typo: parameter*s*

**Review by mattcaswell — APPROVED** [review]
One minor typo. Approved assuming that is fixed.

File: doc/man7/provider-base.pod
Comment (mattcaswell) [reply doc/man7/provider-base.pod]: Exactly. OPENSSL_hexstr2buf is just a utility function. We don't want to bloat the libcrypto<->provider interface with a whole load of utility functions. IMO, we only want to provide the bare essential "OS" level functions that providers couldn't easily provide themselves. If providers really want to take advantage of a broader set of utility functions made available by libcrypto - then they can just link against libcrypto.

Comment (richsalz) [reply doc/man7/provider-base.pod]: EVP requires hexstring conversion.

I did the simplest thing that fixed the build after EVP was brought inside the boundary.  Feel free to improve it, of course.

Comment (mattcaswell) [reply doc/man7/provider-base.pod]: #9432

File: doc/man7/provider-base.pod
Comment (levitte) [reply doc/man7/provider-base.pod]: I'm thinking that I should write "Additionally, provider specific configuration parameters from the config file are available, in dotted form."

If that sounds ok to you, that's what I'll change it to.

Comment (mattcaswell) [reply doc/man7/provider-base.pod]: Sounds fine to me

Comment (levitte) [reply doc/man7/provider-base.pod]: Ok!

**Comment by levitte** [comment]
Merged.

e4c0ec6278f1fbfc50fcdd09769f65ca80866f6b Documentation: Add provider(7), for general description of providers
2cafb1dff3ef13c470c4d2d7b1d8a1f7142d8813 Documentation: Move the description of the fetching functions
d4c69c69d171edb17b4d609c15891a9599809ed0 Documentation: add provider-base(7), describing the base functions

**Pull request closed without merging**
